presto-memory/src/main/java/com/facebook/presto/plugin/memory/MemoryMetadata.java

```diff
@@ -0,0 +1,231 @@
+    {
+        MemoryTableHandle memoryTableHandle = checkType(tableHandle, MemoryTableHandle.class, "tableHandle");
+        return memoryTableHandle.toTableMetadata();
+    }
+
+    @Override
+    public List<SchemaTableName> listTables(ConnectorSession session, String schemaNameOrNull)
+    {
+        if (schemaNameOrNull != null && !schemaNameOrNull.equals(SCHEMA_NAME)) {
+            return ImmutableList.of();
+        }
+        return tables.values().stream()
+                .map(MemoryTableHandle::toSchemaTableName)
+                .collect(toList());
+    }
+
+    @Override
+    public Map<String, ColumnHandle> getColumnHandles(ConnectorSession session, ConnectorTableHandle tableHandle)
+    {
+        MemoryTableHandle memoryTableHandle = checkType(tableHandle, MemoryTableHandle.class, "tableHandle");
+        return memoryTableHandle.getColumnHandles().stream()
+                .collect(toMap(MemoryColumnHandle::getName, Function.identity()));
+    }
+
+    @Override
+    public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTableHandle tableHandle, ColumnHandle columnHandle)
+    {
+        MemoryColumnHandle memoryColumnHandle = checkType(columnHandle, MemoryColumnHandle.class, "columnHandle");
+        return memoryColumnHandle.toColumnMetadata();
+    }
+
+    @Override
+    public Map<SchemaTableName, List<ColumnMetadata>> listTableColumns(ConnectorSession session, SchemaTablePrefix prefix)
+    {
+        return tables.values().stream()
+                .filter(table -> prefix.matches(table.toSchemaTableName()))
+                .collect(toMap(MemoryTableHandle::toSchemaTableName, handle -> handle.toTableMetadata().getColumns()));
+    }
+
+    @Override
+    public void dropTable(ConnectorSession session, ConnectorTableHandle tableHandle)
+    {
+        MemoryTableHandle handle = checkType(tableHandle, MemoryTableHandle.class, "tableHandle");
+        Long tableId = tableIds.remove(handle.getTableName());
+        if (tableId != null) {
+            tables.remove(tableId);
+        }
+    }
+
+    @Override
+    public void renameTable(ConnectorSession session, ConnectorTableHandle tableHandle, SchemaTableName newTableName)
+    {
+        MemoryTableHandle oldTableHandle = checkType(tableHandle, MemoryTableHandle.class, "tableHandle");
+        MemoryTableHandle newTableHandle = new MemoryTableHandle(
+                oldTableHandle.getConnectorId(),
+                oldTableHandle.getSchemaName(),
+                newTableName.getTableName(),
+                oldTableHandle.getTableId(),
+                oldTableHandle.getColumnHandles(),
+                oldTableHandle.getHosts(),
+                oldTableHandle.getActiveTableIds()
+        );
+        tableIds.remove(oldTableHandle.getTableName());
+        tableIds.put(newTableName.getTableName(), oldTableHandle.getTableId());
+        tables.remove(oldTableHandle.getTableId());
+        tables.put(oldTableHandle.getTableId(), newTableHandle);
```

Comment:
Hmmm, good catch. I assumed that `ConnectorMetadata` is used only by single thread.

I will synchronize all methods in `MemoryMetadata`. I think that should solve those problems and indeed synchronization is not an issue here (at least now).